enhancements/authentication/configuring-webhook-token-authenticators.md

```diff
@@ -0,0 +1,176 @@
+
+#### Backend - Integrated OAuth Server Configuration
+
+It's cluster-authentication-operator's responsibility to:
+- observe `authentication.config/cluster` field
+- if set to "IntegratedOAuth", it creates secret `openshift-config/integrated-oauth` and adds it to
+  the `webhookTokenAuthenticators` field of the `authentication.config/cluster` resource
+- the above secret contains details that point kube-apiserver to the oauth-apiserver for token reviews
+
+To perform token reviews for tokens issued by the integrated oauth-server, the new oauth-apiserver needs to
+be able to handle token review requests, such capability is currently missing and needs to be implemented.
+
+When responding after a successful authentication, the oauth-apiserver needs to return all of the users
+groups, on top of which it needs to add the `system:authenticated:oauth` group to keep the behavior of
+previous verions. This group adds self-provisioner cluster role to the authenticated user, the role
+allows users privileges necessary to work with projects.
+
+Similarly, if any integrator would like their users to have project privileges, they need to add this
+specific group in their token review responses.
+
+### Risks and Mitigations
+
+There are a couple of issues that root in the very fact of making it possible of
+replacing the oauth-server with a custom token authenticator:
+
+- `oauth-proxy` relies on the _/.well-known/oauth-authorization-server_ endpoint
+  and if the integrated oauth-server is turned off and the `oauthMetadata` is not set
+  properly, things like grafana and prometheus won't be accessible
+- function of selfubjectaccessreviews and the "users/~" echo endpoint may not work
+  properly
+- ServiceAccounts may not be usable as OAuth clients anymore
+
+It is a sole responsibility of the replacing part to make sure all of these work.
+
+## Design Details
+
+### Test Plan
+
+Since we will be making the cluster-authentication-operator use the `webhookTokenAuthenticatros`
+(in combination with `oauthMetadata`, which already works) to configure authentication using
+the integrated OAuth server, we will just re-use all the tests that we have already.
+
+### Upgrade / Downgrade Strategy
+
+A short period of OpenShift token authentication is perhaps unavoidable as it is
+being torn out of the kube-apiserver process and handed to another one which is
+only going to run after the kube-apiserver pods have been deployed. The period can
+however be shortened by performing the steps necessary to get authentication working back
+again early.
+
+A 4.(n-1) cluster-authentication-operator needs to add the code for adding its own
+secret reference to `webhookTokenAuthenticators` and creating the actual secret in
+`openshift-config` NS. It only does that after the kube-apiserver-operator reaches
+4.n version, so that the operator is capable of processing these artefacts.
```

Comment:
(No action required) TIL that artefact is the british english spelling of artifact

Reply:
I tend to mix British and American English at will to great confusion of the docs people (aside from the mistakes I do based on the differences between Slavic and Germanic languages).